docs: Add net6 and wasm profiling guide (#6986)
unoplatform/uno: doc/articles/guides/profiling-applications.md

# Profiling Uno Platform Applications

## Profiling .NET 6 Android applications

> As of Preview 7, .NET 6 provides the ability to do CPU profiling through [`dotnet-trace`](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-trace) for android applications.

Reviewer: This will open the link in the user preferred language.

```suggestion
As of Preview 7, .NET 6 provides the ability to do CPU profiling through [`dotnet-trace`](https://docs.microsoft.com/dotnet/core/diagnostics/dotnet-trace) for android applications.

```

- When building with AOT, most of the IL is compiled to native code with some exceptions. You may still find `RuntimeHelpers.CompileMethod` invocations. In such cases, you may need to find what is causing the AOT compiler to skip IL portions. If the JIT still impacts cold paths of your application, you may still need to adjust your code to avoid the JIT. For instance, some generics constructs force the AOT compiler to still use JITing. In other cases, it could be accessing static type members. The JIT conditions are runtime version dependent, and [looking at the runtime code](https://github.com/dotnet/runtime/blob/9703660baa08914773b26e413e361c8ce04e6d94/src/mono/mono/mini/aot-compiler.c) can help finding out which ones.
- Some of the time is spent in the .NET Android binding framework (e.g. `Android.Runtime.JNIEnv` or `Java.Interop.TypeManager`), operations that cannot be adjusted by the application. One change to consider is to reduce the native code invocations to the strict minimum, where impactful.

## Profiling Skia/GTK and Skia/WPF applications

> Profiling Skia based Uno Platform targets can be done on Windows in Visual Studio 2019 and 2022 using [time and memory profilers](https://docs.microsoft.com/en-us/visualstudio/profiling/profiling-feature-tour?view=vs-2019).

Reviewer: ```suggestion
Profiling Skia based Uno Platform targets can be done on Windows in Visual Studio 2019 and 2022 using [time and memory profilers](https://docs.microsoft.com/visualstudio/profiling/profiling-feature-tour?view=vs-2019).

```

Author: Thanks for the review!

## Profiling WebAssembly applications

Profiling WebAssembly applications can be done through the use of AOT compilation, and [browsers' performance tab](https://developer.chrome.com/docs/devtools/evaluate-performance/).

### Setup the WebAssembly application for profiling

- Enable emcc profiling:
```xml
<PropertyGroup>
    <WasmShellEnableEmccProfiling>true</WasmShellEnableEmccProfiling>
</PropertyGroup>
```
- Enable AOT compilation:
```xml
<PropertyGroup>
    <WasmShellEnableEmccProfiling>InterpreterAndAOT</WasmShellEnableEmccProfiling>
</PropertyGroup>

> ```

Reviewer: ~~~suggestion
- Enable emcc profiling:
    ```xml
    <PropertyGroup>
        <WasmShellEnableEmccProfiling>true</WasmShellEnableEmccProfiling>
    </PropertyGroup>
    ```
- Enable AOT compilation:
    ```xml
    <PropertyGroup>
        <WasmShellEnableEmccProfiling>InterpreterAndAOT</WasmShellEnableEmccProfiling>
    </PropertyGroup>
    ```

~~~

    ```

Note that the `suspend` directive means that if `dotnet-trace` is not running, the application waits for it to start.

### Profiling the application

- Start the diagnostics router, in any folder:
    ```
    dotnet-dsrouter client-server -tcps 127.0.0.1:9001 -ipcc /tmp/uno-app --verbose debug
    ```
- Start `dotnet-trace`, in the app folder or where you want your traces to be stored:
    ```
    dotnet-trace collect --diagnostic-port /tmp/uno-app --format speedscope -o uno-app-trace
    ```
- Start an `x86-64` emulator or `arm64` (`armv8`) device
    > Running on a 32 bits device is not supported and will generate unusable traces in SpeedScope
- Build the application with profiling enabled
    ```
    dotnet build -f net6.0-android -t:run -c Release -p:IsEmulator=true /p:RunAOTCompilation=true /p:AndroidEnableProfiler=true
    ```
- The app will start and the `dotnet-trace` will display a MB number counting up
- Use the app and once done, stop `dotnet-trace` using the specified method (Likely `Enter` or `Ctr+C`)
- Open a browser at `https://speedscope.app` and drop the `uno-app-trace.speedscope.json` file on it

### Analyzing the trace data

> This section is meant to be expanding with time, providing insights into what to look for when analyzing flame charts.

Reviewer 2: ```suggestion
This section provides insights into what to look for when analyzing flame charts.

```
